### PR TITLE
Add akv token request logic

### PR DESCRIPTION
**Pull request by Xtrimmer**
In the last Iteration we added support for Always Encrypted with secrets stored as local certificates. This change adds support for secrets stored in Azure Key Vault.

In the event that a user executes a select query, and one or more columns are encrypted using the Always Encrypted feature, and where the decryption key exists in AKV, then an access token will be requested under the authority of a Linked Azure Account and provided to the Microsoft.Data.SqlClient Driver so that it can acquire the key from AKV and use it to decrypt the query results data. 

This is half of the solution with the other half of the [changes made in SqlToolsService](https://github.com/microsoft/sqltoolsservice/pull/934).

### CODE REVIEW

File: extensions/mssql/src/features.ts [hunk 1]
```diff
@@ -27,6 +27,41 @@ export class TelemetryFeature implements StaticFeature {
...
+		Utils.ensure(capabilities, 'telemetry')!.telemetry = true;
+	}
+
+	initialize(): void {
+		this._client.onRequest(contracts.SecurityTokenRequest.type, async e => {
+			const accountList = await azdata.accounts.getAllAccounts();
+
+			if (accountList.length < 1) {
+				// TODO: Prompt user to add account
+				window.showErrorMessage('Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, but no linked Azure account is available. Please add a linked Azure account and retry the query.');
```
Comment (abist) [inline extensions/mssql/src/features.ts]: These messages should be localized

Comment (aaomidi) [reply extensions/mssql/src/features.ts]: Example https://github.com/microsoft/azuredatastudio/blob/4a68ab46590134be0fa663cb59ac59b8603df1d1/extensions/admin-tool-ext-win/src/main.ts#L114

File: extensions/mssql/src/contracts.ts [hunk 2]
```diff
@@ -31,6 +31,22 @@ export class TelemetryParams {
 
 // ------------------------------- </ Telemetry Sent Event > ----------------------------------
 
+// ------------------------------- < Security Token Request > ------------------------------------------
+export interface RequestSecurityTokenParams {
+	authority: string;
```
Comment (aaomidi) [inline extensions/mssql/src/contracts.ts]: What is authority here?

Comment (Xtrimmer) [reply extensions/mssql/src/contracts.ts]: The authority is the Identifier of the authority, a URL. It (and those other members) maps to the [KeyVaultClient.AuthenticationCallback Delegate](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.keyvault.keyvaultclient.authenticationcallback?view=azure-dotnet-legacy) parameters.
An example of authority would be:
`{"authority": "https://login.windows.net/XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"}`

File: extensions/mssql/src/features.ts (hunk 1 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Fixed.

File: extensions/mssql/src/contracts.ts (hunk 2 above)
Comment (aaomidi) [reply extensions/mssql/src/contracts.ts]: I think you should add Provider as well - so we know when this comes in is it a request for an azure account or {whateverService} account.

File: extensions/mssql/src/features.ts [hunk 3]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
...
+		Utils.ensure(capabilities, 'telemetry')!.telemetry = true;
+	}
+
+	initialize(): void {
+		this._client.onRequest(contracts.SecurityTokenRequest.type, async e => {
+			const accountList = await azdata.accounts.getAllAccounts();
+
+			if (accountList.length < 1) {
+				// TODO: Prompt user to add account
+				window.showErrorMessage(localize('mssql.missingLinkedAzureAccount', 'Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, but no linked Azure account is available. Please add a linked Azure account and retry the query.'));
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: The second parameter should be quoted with " instead of '

File: extensions/mssql/src/features.ts [hunk 4]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
...
+				return null;
+			} else if (accountList.length > 1) {
+				// TODO: Prompt user to select an account
+				window.showErrorMessage(localize('mssql.multipleLinkedAzureAccount', 'Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, which is not supported if multiple linked Azure accounts are present. Make sure only one linked Azure account exists and retry the query.'));
+				return null;
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: Use undefined instead of null

File: extensions/mssql/src/features.ts [hunk 5]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
 	}
 }
 
+export class AccountFeature implements StaticFeature {
+
+	constructor(private _client: SqlOpsDataClient) { }
+
+	fillClientCapabilities(capabilities: ClientCapabilities): void {
+		Utils.ensure(capabilities, 'telemetry')!.telemetry = true;
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: Not necessary

File: extensions/mssql/src/features.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply extensions/mssql/src/features.ts]: And what is the behavior if we return undefined here? How does the tools service side handle that?

File: extensions/mssql/src/features.ts [hunk 6]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
...
+				window.showErrorMessage(localize('mssql.multipleLinkedAzureAccount', 'Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, which is not supported if multiple linked Azure accounts are present. Make sure only one linked Azure account exists and retry the query.'));
+				return null;
+			}
+
+			let account = accountList[0];
+			const securityToken: { [key: string]: any } = await azdata.accounts.getSecurityToken(account, azdata.AzureResource.AzureKeyVault);
+			const tenant = account.properties.tenants.find((t: { [key: string]: string }) => e.authority.includes(t.id));
+
+			let params: contracts.RequestSecurityTokenResponse = {
+				token: securityToken[tenant.id].token
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: tenant could be undefined here

File: extensions/mssql/src/features.ts [hunk 7]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
...
+				window.showErrorMessage(localize('mssql.missingLinkedAzureAccount', 'Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, but no linked Azure account is available. Please add a linked Azure account and retry the query.'));
+				return null;
+			} else if (accountList.length > 1) {
+				// TODO: Prompt user to select an account
+				window.showErrorMessage(localize('mssql.multipleLinkedAzureAccount', 'Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, which is not supported if multiple linked Azure accounts are present. Make sure only one linked Azure account exists and retry the query.'));
+				return null;
+			}
+
+			let account = accountList[0];
+			const securityToken: { [key: string]: any } = await azdata.accounts.getSecurityToken(account, azdata.AzureResource.AzureKeyVault);
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: Don't you need to update https://github.com/microsoft/azuredatastudio/blob/master/extensions/azurecore/src/account-provider/azureAccountProvider.ts#L159 as well?

Has this experience been tested end-to-end successfully?

File: extensions/mssql/src/features.ts (hunk 6 above)
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: securityToken might not contain the tenant we're looking for so accessing the token property here isn't safe

File: extensions/azurecore/src/account-provider/azureAccountProvider2.ts [hunk 8]
```diff
@@ -92,7 +92,8 @@ export class AzureAccountProvider implements azdata.AccountProvider {
 		const resourceIdMap = new Map<azdata.AzureResource, string>([
 			[azdata.AzureResource.ResourceManagement, this.metadata.settings.armResource.id],
 			[azdata.AzureResource.Sql, this.metadata.settings.sqlResource.id],
-			[azdata.AzureResource.OssRdbms, this.metadata.settings.ossRdbmsResource.id]
+			[azdata.AzureResource.OssRdbms, this.metadata.settings.ossRdbmsResource.id],
+			[azdata.AzureResource.AzureKeyVault, this.metadata.settings.azureKeyVaultResource.id]
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/account-provider/azureAccountProvider2.ts]: I believe you'll want to update https://github.com/microsoft/azuredatastudio/blob/master/extensions/azurecore/src/account-provider/azureAccountProvider.ts#L159 too

File: src/sql/azdata.proposed.d.ts [hunk 9]
```diff
@@ -121,7 +121,8 @@ declare module 'azdata' {
 	 * Add OssRdbms for sqlops AzureResource.
 	 */
 	export enum AzureResource {
-		OssRdbms = 2
+		OssRdbms = 2,
```
Comment (abist) [inline src/sql/azdata.proposed.d.ts]: This is not really related to your PR and I may be missing context here, but why is this enum different from the one in interfaces? This seems to be missing `Sql` and `ResourceManagement`. @aaomidi any idea?

Comment (Charles-Gagnon) [reply src/sql/azdata.proposed.d.ts]: Because this is the newly-added ones to the public api so are in the proposed typings file. The rest are in https://github.com/microsoft/azuredatastudio/blob/master/src/sql/azdata.d.ts#L2210

Comment (abist) [reply src/sql/azdata.proposed.d.ts]: Isn't proposed supposed to be a super set of the actual API? Or do we only add the new additions here with the actual API as a fallback?

Comment (Charles-Gagnon) [reply src/sql/azdata.proposed.d.ts]: What do you mean fallback? Proposed is where new additions to the API go to be tested - things in proposed are able to have breaking changes made to them if we want to. This is the contract we've adopted from VS Code.

https://code.visualstudio.com/api/advanced-topics/using-proposed-api

File: extensions/mssql/src/features.ts (hunk 5 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Oh yes, Ctrl+P Ctrl+V and forgot to remove. Thanks!

File: extensions/azurecore/src/account-provider/azureAccountProvider2.ts (hunk 8 above)
Comment (Xtrimmer) [reply extensions/azurecore/src/account-provider/azureAccountProvider2.ts]: Done.

File: extensions/mssql/src/features.ts (hunk 3 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Fixed.

File: extensions/mssql/src/features.ts [hunk 10]
```diff
@@ -27,6 +29,41 @@ export class TelemetryFeature implements StaticFeature {
...
+				window.showErrorMessage(localize('mssql.multipleLinkedAzureAccount', "Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, which is not supported if multiple linked Azure accounts are present. Make sure only one linked Azure account exists and retry the query."));
+				return undefined;
+			}
+
+			let account = accountList[0];
+			const securityToken: { [key: string]: any } = await azdata.accounts.getSecurityToken(account, azdata.AzureResource.AzureKeyVault);
+			const tenant = account.properties.tenants.find((t: { [key: string]: string }) => e.authority.includes(t.id));
+
+			let params: contracts.RequestSecurityTokenResponse = {
+				token: securityToken[tenant.id].token
```
Comment (aaomidi) [inline extensions/mssql/src/features.ts]: Would it also make sense for you to send the accountKey over? If that accountKey has succeeded you can send it back in future requests when you need a new token that is refreshed, etc.

File: extensions/mssql/src/features.ts (hunk 4 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Currently SqlToolsService ignores it. On the front end the user is (termorarily) presented with the message to add an account or use a single account. If not then the query is stopped and is the same as if the query is canceled by the user. We will be adding more for an improved UX in following iterations, which is what the TODOs are denoting.

Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Changed null to undefined as requested.

File: extensions/mssql/src/contracts.ts (hunk 2 above)
Comment (Xtrimmer) [reply extensions/mssql/src/contracts.ts]: Provider added.

File: extensions/mssql/src/features.ts (hunk 6 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Added undefined check.

Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Added undefined check.

File: extensions/mssql/src/features.ts (hunk 10 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Added account key to response object.

File: extensions/mssql/src/features.ts (hunk 5 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Left the function body empty except for a comment noting that implementation is not necessary.

File: extensions/mssql/src/features.ts [hunk 11]
```diff
@@ -27,6 +29,52 @@ export class TelemetryFeature implements StaticFeature {
...
+export class AccountFeature implements StaticFeature {
+
+	constructor(private _client: SqlOpsDataClient) { }
+
+	fillClientCapabilities(capabilities: ClientCapabilities): void {
+		// this isn't explicitly necessary
+	}
+
+	initialize(): void {
+		this._client.onRequest(contracts.SecurityTokenRequest.type, async e => {
```
Comment (aaomidi) [inline extensions/mssql/src/features.ts]: Just so it's clear, can you have the return type of this function here? `Promise<contracts.RequestSecurityTokenResponse | undefined>`

File: extensions/mssql/src/features.ts [hunk 12]
```diff
@@ -27,6 +29,52 @@ export class TelemetryFeature implements StaticFeature {
 	}
 }
 
+export class AccountFeature implements StaticFeature {
+
+	constructor(private _client: SqlOpsDataClient) { }
+
+	fillClientCapabilities(capabilities: ClientCapabilities): void {
+		// this isn't explicitly necessary
```
Comment (aaomidi) [inline extensions/mssql/src/features.ts]: Remove this comment

File: extensions/mssql/src/features.ts [hunk 13]
```diff
@@ -27,6 +29,52 @@ export class TelemetryFeature implements StaticFeature {
...
+	}
+
+	initialize(): void {
```
Comment (aaomidi) [inline extensions/mssql/src/features.ts]: @Charles-Gagnon do we want a telemetry notification when we get an AE request? Would it help to see adoption?

File: extensions/mssql/src/features.ts [hunk 14]
```diff
@@ -27,6 +29,52 @@ export class TelemetryFeature implements StaticFeature {
...
+				// TODO: Prompt user to select an account
+				window.showErrorMessage(localize('mssql.multipleLinkedAzureAccount', "Azure Data Studio needs to contact Azure Key Vault to access a column master key for Always Encrypted, which is not supported if multiple linked Azure accounts are present. Make sure only one linked Azure account exists and retry the query."));
+				return undefined;
+			}
+
+			let account = accountList[0];
+			const securityToken: { [key: string]: any } = await azdata.accounts.getSecurityToken(account, azdata.AzureResource.AzureKeyVault);
+			const tenant = account.properties.tenants.find((t: { [key: string]: string }) => e.authority.includes(t.id));
+			const unauthorizedMessage = localize('mssql.insufficientlyPrivelagedAzureAccount', "The configured Azure account for {0} does not have sufficient permissions for Azure Key Vault to access a column master key for Always Encrypted.", account.key.accountId);
+			if (tenant === undefined) {
```
Comment (aaomidi) [inline extensions/mssql/src/features.ts]: Do `!tenant` instead of `tenant === undefined`. The `=== undefined` check doesn't check for it potentially being null.

Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Fixed.

File: extensions/mssql/src/features.ts (hunk 12 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Removed.

File: extensions/mssql/src/features.ts (hunk 11 above)
Comment (Xtrimmer) [reply extensions/mssql/src/features.ts]: Type def added.

File: extensions/azurecore/src/account-provider/azureAccountProvider2.ts [hunk 15]
```diff
@@ -96,7 +96,8 @@ export class AzureAccountProvider implements azdata.AccountProvider {
...
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/account-provider/azureAccountProvider2.ts]: @aaomidi Your PR is deleting these files - I assume you'll handle making sure this is moved over correctly once this is merged in?

File: extensions/mssql/src/features.ts [hunk 16]
```diff
@@ -27,6 +29,50 @@ export class TelemetryFeature implements StaticFeature {
...
+	constructor(private _client: SqlOpsDataClient) { }
+
+	fillClientCapabilities(capabilities: ClientCapabilities): void { }
+
+	initialize(): void {
+		this._client.onRequest(contracts.SecurityTokenRequest.type, async (e): Promise<contracts.RequestSecurityTokenResponse | undefined> => {
+			const accountList = await azdata.accounts.getAllAccounts();
+
+			if (accountList.length < 1) {
+				// TODO: Prompt user to add account
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/features.ts]: Please make sure to make issues for tracking these.

File: extensions/mssql/src/features.ts (hunk 13 above)
Comment (Charles-Gagnon) [reply extensions/mssql/src/features.ts]: Yeah good call, it looks like this is using the older telemetry library though so once this is merged in I'll go through and update that and add the events in.
